### ISSUE TITLE
ETIMEDOUT 40.83.177.42:8883 using azure-iot-device-mqtt 

### ISSUE THREAD

**Issue by albhardy** [Azure/azure-iot-sdk-node#6]
# Context
- **OS and version used:** Windows 10
- **Node.js version:** v6.4.0
- **npm version:** 3.10.3
- **list of installed packages:** 

      simulateddevice@1.0.0 
      D:\DEV\AZURE\IOTHUB\simulateddevice
        +-- azure-iot-device@1.1.2
        | +-- azure-iot-common@1.1.2
        | | `-- crypto@0.0.3
        | +-- azure-iot-http-base@1.1.2
        | +-- azure-storage@1.4.0
        | | +-- browserify-mime@1.2.9
        | | +-- extend@1.2.1
        | | +-- json-edm-parser@0.1.2
        | | | `-- jsonparse@1.2.0
        | | +-- node-uuid@1.4.7
        | | +-- readable-stream@2.0.6
        | | | +-- core-util-is@1.0.2
        | | | +-- isarray@1.0.0
        | | | +-- process-nextick-args@1.0.7
        | | | +-- string_decoder@0.10.31
        | | | `-- util-deprecate@1.0.2
        | | +-- request@2.74.0
        | | | +-- aws-sign2@0.6.0
        | | | +-- aws4@1.5.0
        | | | +-- bl@1.1.2
        | | | +-- caseless@0.11.0
        | | | +-- combined-stream@1.0.5
        | | | | `-- delayed-stream@1.0.0
        | | | +-- extend@3.0.0
        | | | +-- forever-agent@0.6.1
        | | | +-- form-data@1.0.1
        | | | | `-- async@2.1.4
        | | | +-- har-validator@2.0.6
        | | | | +-- chalk@1.1.3
        | | | | | +-- ansi-styles@2.2.1
        | | | | | +-- escape-string-regexp@1.0.5
        | | | | | +-- has-ansi@2.0.0
        | | | | | | `-- ansi-regex@2.0.0
        | | | | | +-- strip-ansi@3.0.1
        | | | | | `-- supports-color@2.0.0
        | | | | +-- commander@2.9.0
        | | | | | `-- graceful-readlink@1.0.1
        | | | | +-- is-my-json-valid@2.15.0
        | | | | | +-- generate-function@2.0.0
        | | | | | +-- generate-object-property@1.2.0
        | | | | | | `-- is-property@1.0.2
        | | | | | `-- jsonpointer@4.0.0
        | | | | `-- pinkie-promise@2.0.1
        | | | |   `-- pinkie@2.0.4
        | | | +-- hawk@3.1.3
        | | | | +-- boom@2.10.1
        | | | | +-- cryptiles@2.0.5
        | | | | +-- hoek@2.16.3
        | | | | `-- sntp@1.0.9
        | | | +-- http-signature@1.1.1
        | | | | +-- assert-plus@0.2.0
        | | | | +-- jsprim@1.3.1
        | | | | | +-- extsprintf@1.0.2
        | | | | | +-- json-schema@0.2.3
        | | | | | `-- verror@1.3.6
        | | | | `-- sshpk@1.10.1
        | | | |   +-- asn1@0.2.3
        | | | |   +-- assert-plus@1.0.0
        | | | |   +-- bcrypt-pbkdf@1.0.0
        | | | |   +-- dashdash@1.14.1
        | | | |   | `-- assert-plus@1.0.0
        | | | |   +-- ecc-jsbn@0.1.1
        | | | |   +-- getpass@0.1.6
        | | | |   | `-- assert-plus@1.0.0
        | | | |   +-- jodid25519@1.0.2
        | | | |   +-- jsbn@0.1.0
        | | | |   `-- tweetnacl@0.14.5
        | | | +-- is-typedarray@1.0.0
        | | | +-- isstream@0.1.2
        | | | +-- json-stringify-safe@5.0.1
        | | | +-- mime-types@2.1.13
        | | | | `-- mime-db@1.25.0
        | | | +-- oauth-sign@0.8.2
        | | | +-- qs@6.2.1
        | | | +-- stringstream@0.0.5
        | | | +-- tough-cookie@2.3.2
        | | | | `-- punycode@1.4.1
        | | | `-- tunnel-agent@0.4.3
        | | +-- underscore@1.4.4
        | | +-- validator@3.22.2
        | | +-- xml2js@0.2.7
        | | | `-- sax@0.5.2
        | | `-- xmlbuilder@0.4.3
        | +-- debug@2.4.5
        | | `-- ms@0.7.2
        | +-- es5-shim@4.5.9
        | +-- lodash@4.17.2
        | +-- machina@2.0.0
        | | `-- lodash@3.10.1
        | `-- traverse@0.6.6
        +-- azure-iot-device-amqp@1.1.2
        | `-- azure-iot-amqp-base@1.1.2
        |   +-- amqp10@3.3.2
        |   | +-- bluebird@3.4.6
        |   | +-- buffer-builder@0.2.0
        |   | +-- node-amqp-encoder@0.0.2
        |   | +-- node-int64@0.4.0
        |   | +-- stately.js@1.3.0
        |   | `-- uuid@2.0.3
        |   +-- amqp10-transport-ws@0.0.5
        |   | `-- nodejs-websocket@1.7.1
        |   +-- sinon@1.17.6
        |   | +-- formatio@1.1.1
        |   | +-- lolex@1.3.2
        |   | +-- samsam@1.1.2
        |   | `-- util@0.10.3
        |   |   `-- inherits@2.0.1
        |   `-- sinon-as-promised@4.0.2
        |     +-- create-thenable@1.0.2
        |     | +-- object.omit@2.0.1
        |     | | +-- for-own@0.1.4
        |     | | | `-- for-in@0.1.6
        |     | | `-- is-extendable@0.1.1
        |     | `-- unique-concat@0.2.2
        |     `-- native-promise-only@0.8.1
        +-- azure-iot-device-http@1.1.2
        | `-- node-crontab@0.0.8
        |   `-- cron-parser@0.3.6
        `-- azure-iot-device-mqtt@1.1.2
          `-- mqtt@1.14.1
            +-- commist@1.0.0
            | `-- leven@1.0.2
            +-- concat-stream@1.6.0
            | +-- readable-stream@2.2.2
            | | `-- buffer-shims@1.0.0
            | `-- typedarray@0.0.6
            +-- end-of-stream@1.1.0
            | `-- once@1.3.3
            |   `-- wrappy@1.0.2
            +-- help-me@1.0.1
            | +-- callback-stream@1.1.0
            | +-- glob-stream@5.3.5
            | | +-- extend@3.0.0
            | | +-- glob@5.0.15
            | | | +-- inflight@1.0.6
            | | | +-- minimatch@3.0.3
            | | | | `-- brace-expansion@1.1.6
            | | | |   +-- balanced-match@0.4.2
            | | | |   `-- concat-map@0.0.1
            | | | `-- path-is-absolute@1.0.1
            | | +-- glob-parent@3.1.0
            | | | +-- is-glob@3.1.0
            | | | | `-- is-extglob@2.1.1
            | | | `-- path-dirname@1.0.2
            | | +-- micromatch@2.3.11
            | | | +-- arr-diff@2.0.0
            | | | | `-- arr-flatten@1.0.1
            | | | +-- array-unique@0.2.1
            | | | +-- braces@1.8.5
            | | | | +-- expand-range@1.8.2
            | | | | | `-- fill-range@2.2.3
            | | | | |   +-- is-number@2.1.0
            | | | | |   +-- isobject@2.1.0
            | | | | |   +-- randomatic@1.1.6
            | | | | |   `-- repeat-string@1.6.1
            | | | | +-- preserve@0.2.0
            | | | | `-- repeat-element@1.1.2
            | | | +-- expand-brackets@0.1.5
            | | | | `-- is-posix-bracket@0.1.1
            | | | +-- extglob@0.3.2
            | | | | `-- is-extglob@1.0.0
            | | | +-- filename-regex@2.0.0
            | | | +-- is-extglob@1.0.0
            | | | +-- is-glob@2.0.1
            | | | +-- kind-of@3.1.0
            | | | | `-- is-buffer@1.1.4
            | | | +-- normalize-path@2.0.1
            | | | +-- parse-glob@3.0.4
            | | | | +-- glob-base@0.3.0
            | | | | | +-- glob-parent@2.0.0
            | | | | | `-- is-glob@2.0.1
            | | | | |   `-- is-extglob@1.0.0
            | | | | +-- is-dotfile@1.0.2
            | | | | +-- is-extglob@1.0.0
            | | | | `-- is-glob@2.0.1
            | | | `-- regex-cache@0.4.3
            | | |   +-- is-equal-shallow@0.1.3
            | | |   `-- is-primitive@2.0.0
            | | +-- ordered-read-streams@0.3.0
            | | | `-- is-stream@1.1.0
            | | +-- through2@0.6.5
            | | | `-- readable-stream@1.0.34
            | | |   `-- isarray@0.0.1
            | | +-- to-absolute-glob@0.1.1
            | | | `-- extend-shallow@2.0.1
            | | `-- unique-stream@2.2.1
            | |   +-- json-stable-stringify@1.0.1
            | |   | `-- jsonify@0.0.0
            | |   `-- through2-filter@2.0.0
            | `-- through2@2.0.3
            |   `-- readable-stream@2.2.2
            +-- inherits@2.0.3
            +-- minimist@1.2.0
            +-- mqtt-connection@2.1.1
            | +-- reduplexer@1.1.0
            | | `-- readable-stream@1.0.34
            | |   `-- isarray@0.0.1
            | `-- through2@0.6.5
            |   `-- readable-stream@1.0.34
            |     `-- isarray@0.0.1
            +-- mqtt-packet@3.4.7
            | `-- bl@0.9.5
            |   `-- readable-stream@1.0.34
            |     `-- isarray@0.0.1
            +-- pump@1.0.2
            +-- readable-stream@1.0.34
            | `-- isarray@0.0.1
            +-- reinterval@1.1.0
            +-- split2@2.1.0
            +-- websocket-stream@3.3.3
            | +-- duplexify@3.5.0
            | | +-- end-of-stream@1.0.0
            | | `-- stream-shift@1.0.0
            | `-- ws@1.1.1
            |   +-- options@0.0.6
            |   `-- ultron@1.0.2
            `-- xtend@4.0.1
   

# Code sample exhibiting the issue: 
I run [simple_sample_device_with_sas.js](https://github.com/Azure/azure-iot-sdk-node/blob/master/device/samples/simple_sample_device_with_sas.js) example and experienced following protocols are not working:
- var Protocol = require('azure-iot-device-mqtt').Mqtt results in `Could not connect: connect ETIMEDOUT 40.83.177.42:8883` 
- var Protocol = require('azure-iot-device-amqp').Amqp; results in `Could not connect: Error: AMQP Transport: Could not connect`

However, these protocols work:
- var Protocol = require('azure-iot-device-amqp').AmqpWs;
- var Protocol = require('azure-iot-device-http').Http;

Is there anything else that needs to be configured to enable send/receive MQTT on Windows? I also noticed `Device Explorer` is also unable to receive message from IOT Hub:

![image](https://cloud.githubusercontent.com/assets/1732018/21381874/bea9db8c-c798-11e6-96a1-07bf971eaea9.png)

**Comment by olivierbloch** [comment]
Hi @albhardy
All the symptoms you report seem to indicate that you are running behind a firewall that blocks the amqp amd mqtt protocols ports.
You can use this tool to confirm:
https://github.com/Azure/iothub-diagnostics
Note that ports being blocked by firewalls is the main reason why we also support these protocols (amqp and mqtt) over WebSockets.

**Comment by albhardy** [comment]
I suspect that - the app works in my home environment. I could not use iothub-diagnostic to check ([due to node-gyp rebuild issue](https://github.com/Azure/iothub-diagnostics/issues/1)); 

However, you mentioned amqp support over WebSockets, any idea how to configure `var EventHubClient = require('azure-event-hubs').Client;` in [ReadDeviceToCloud.js example](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-node-node-getstarted) to read via amqp WebSocket? I could not find documentation on this yet. It's also not clear from [Azure-Event-Hub](https://www.npmjs.com/package/azure-event-hubs) npm page.

**Comment by anhashia** [comment]
Investigating !

**Comment by tameraw** [comment]
This issue was moved to Azure/azure-event-hubs-dotnet#88

**Comment by tameraw** [comment]
Correction. This was moved to the "node" repo for event hub instead.
New issue can be found here:
https://github.com/Azure/azure-event-hubs-node/issues/12
